src/renderers/shared/stack/reconciler/ReactRef.js

```diff
@@ -56,20 +66,21 @@ ReactRef.shouldUpdateRefs = function(prevElement, nextElement) {
   // is made. It probably belongs where the key checking and
   // instantiateReactComponent is done.
 
-  var prevEmpty = prevElement === null || prevElement === false;
-  var nextEmpty = nextElement === null || nextElement === false;
-
   return (
     // This has a few false positives w/r/t empty components.
-    prevEmpty || nextEmpty ||
+    prevElement === null || prevElement === false || // prevEmpty
+    nextElement === null || nextElement === false || // nextEmpty
     nextElement.ref !== prevElement.ref ||
     // If owner changes but we have an unchanged function ref, don't update refs
     (typeof nextElement.ref === 'string' &&
      nextElement._owner !== prevElement._owner)
   );
 };
 
-ReactRef.detachRefs = function(instance, element) {
+ReactRef.detachRefs = function(
+  instance: ReactInstance,
+  element: ReactElement,
```

Comment:
Good catch, I can't wait for flow to warn when you're doing checks that do not make sense based on the type you have.